src/EventProducerService.cpp

```diff
@@ -78,43 +82,28 @@ void EventProducerService::sendEvent(bool state, byte evValue)
       nn_en[0] = highByte(module_config->nodeNum);
       nn_en[1] = lowByte(module_config->nodeNum); 
     }
-    else if ((nn_en[0] == 0xff) && (nn_en[1] == 0xff))
-    {
-      createDefaultEvent(evValue);
-      return;
-    }
     else
     {
+      if ((nn_en[0] == 0xff) && (nn_en[1] == 0xff))
+      {
+        createDefaultEvent(evValue);
+      }    
       opCode = (state ? OPC_ACON : OPC_ACOF);
     }
-    
-    VlcbMessage msg;
-    msg.len = 5;
-    msg.data[0] = opCode;
-    msg.data[1] = nn_en[0];
-    msg.data[2] = nn_en[1];
-    msg.data[3] = nn_en[2];
-    msg.data[4] = nn_en[3];
-    controller->sendMessage(&msg);
-      
-    if (coeService)
-    {
-      coeService->put(&msg);
-    }
-  }
-  else
-  {
-    createDefaultEvent(evValue);
   }
+  generateMessage();  
```

Comment:
Hmm, this doesn't look right. What if the if-statement on line 71 is true? Then this line will generate a message where opcode and nn_en are not set.
If you used function parameters instead of class variables then this would be easier to spot.

Reply:
nn_en will have been set by the createDefaultEvent(), but you're right about the op-code being undefined.

Reply:
Oh, relying on such side effects is scary.

Reply:
It will be set at line 91

Reply:
This actually turned out to require more thinking than I'd expected.  Moving the declaration of nn_en and opCode to the .h file so that they stayed in scope across a number of methods made it easier.  I can think further about it and use arguments in the calls instead if that gives better code.

Reply:
Line 91 is in the false branch of the if-statement on line 71, thus it will not be executed if that condition is true.

Reply:
No it isn't.  It is always executed if the nn != 0xFFFF

Reply:
The else branch for the if statement on line 71 spans lines 76-92. It is tricky with all the nested if statements but as I read the code line 91 will not be executed if the condition on line 71 is false.

Reply:
Ah, yes, I see what you mean.  I've duplicated line 91 after line 73 to resolve.  I can't move line 91 or I'll upset the short event op-codes.